### ISSUE TITLE
Port twisted.plugin to Python3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7182]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Alex reported|
|-|-|
|Trac ID|trac#7182|
|Type|enhancement|
|Created|2014-04-16 16:02:18Z|
|Branch|https://github.com/twisted/twisted/tree/py3-plugin-7182|

<details><summary>Searchable metadata</summary>

```
trac-id__7182 7182
type__enhancement enhancement
reporter__Alex Alex
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_py3_plugin_7182 branches/py3-plugin-7182
branch_author__mesozoic mesozoic
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1397664138000000 1397664138000000
changetime__1431677455632650 1431677455632650
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|Alex commented|
|-|-|

This is resolved by https://github.com/twisted/twisted/pull/22

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|Alex commented|
|-|-|

Err -- there's a bunch of failures that I missed somehow, not ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_3"></a>|@mesozoic commented|
|-|-|

(In [44695]) Branching to py3-plugin-7182.

**Comment by twisted-trac** [comment]
|[<img alt="mesozoic's avatar" src="https://avatars.githubusercontent.com/u/458347?s=50" width="50" height="50">](https://github.com/mesozoic)<a name="note_4"></a>|@mesozoic commented|
|-|-|

We are running into differences in import behavior between 2.7 and 3.4. Any modifications we make to the application-provided plugins at runtime are not getting picked up by `import`.

```python
def test_freshPyReplacesStalePyc(self):
    mypath = self.appPackage.child("stale.py")
    mypath.setContent(pluginFileContents('one'))
    from plugindummy.plugins import stale  # this fails on py3, but not on py2
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Looks like we need to call [importlib.invalidate_caches](https://docs.python.org/3.4/library/importlib.html#importlib.invalidate_caches).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph removed owner|
|-|-|

I think this is ready for re-review now; all the tests on py3 pass.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Builders are still spinning, but should be mostly done shortly.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl set owner to @glyph|
|-|-|

mesozoic, glyph -- thanks for the patch.

- twisted/plugin.py: in ```twisted.python.compat``` there is an iteritems function that returns an iterator in both Python2 and Python3. Changing it to just ```.items()``` makes it less efficient on Python3.
- All Py3-compatible files need ```from __future__ import absolute_import, division``` added to them.
- The output of the repr isn't covered. I think it should work fine in Python 3, and I'm not sure bytestrings can get in there, but a test would be nice.
- The newsfile should be "twisted.plugin now supports Python 3", in line with the newsfile standard.

Otherwise, it looks good. Please fix the above issues and merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [hawkowl](#note_8):
> mesozoic, glyph -- thanks for the patch.
> 
> - twisted/plugin.py: in ```twisted.python.compat``` there is an iteritems function that returns an iterator in both Python2 and Python3. Changing it to just ```.items()``` makes it less efficient on Python3.
> - All Py3-compatible files need ```from __future__ import absolute_import, division``` added to them.
> - The output of the repr isn't covered. I think it should work fine in Python 3, and I'm not sure bytestrings can get in there, but a test would be nice.
> - The newsfile should be "twisted.plugin now supports Python 3", in line with the newsfile standard.
> 
> Otherwise, it looks good. Please fix the above issues and merge.

OK, issues should be fixed, but running the builders to make sure.

**Comment by twisted-trac** [comment]
|<img alt="adiroiban1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|adiroiban1 commented|
|-|-|

Changes look good. Happy to see it merged as I bumped into this while working on porting Trial :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

I'm happy with this, so I'm going to merge it for you, Glyph :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl commented|
|-|-|

Also, I reviewed this, whoops.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44820]) Merge py3-plugin-7182: Port twisted.plugin to Python 3

Author: mesozoic, glyph
Reviewer: hawkowl
Fixes: [#7182](https://github.com/twisted/twisted/issues/7182)
